miner/scroll_worker.go

```diff
@@ -0,0 +1,945 @@
+		log.Error("failed to start pipeline", "err", err)
+		return
+	}
+
+	// Short circuit if there is no available pending transactions.
+	// But if we disable empty precommit already, ignore it. Since
+	// empty block is necessary to keep the liveness of the network.
+	if len(localTxs) == 0 && len(remoteTxs) == 0 && len(l1Messages) == 0 && atomic.LoadUint32(&w.noempty) == 0 {
+		return
+	}
+
+	if w.chainConfig.Scroll.ShouldIncludeL1Messages() && len(l1Messages) > 0 {
+		log.Trace("Processing L1 messages for inclusion", "count", len(l1Messages))
+		txs, err := types.NewL1MessagesByQueueIndex(l1Messages)
+		if err != nil {
+			log.Error("Failed to create L1 message set", "l1Messages", l1Messages, "err", err)
+			return
+		}
+
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
+			w.handleCCCResult(result)
+			return
+		}
+	}
+	signer := types.MakeSigner(w.chainConfig, header.Number)
+
+	if w.prioritizedTx != nil && w.currentPipeline.Header.Number.Uint64() > w.prioritizedTx.blockNumber {
+		w.prioritizedTx = nil
+	}
+	if w.prioritizedTx != nil {
+		from, _ := types.Sender(signer, w.prioritizedTx.tx) // error already checked before
+		txList := map[common.Address]types.Transactions{from: []*types.Transaction{w.prioritizedTx.tx}}
+		txs := types.NewTransactionsByPriceAndNonce(signer, txList, header.BaseFee)
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
+			w.handleCCCResult(result)
+			return
+		}
+	}
+
+	if len(localTxs) > 0 {
+		txs := types.NewTransactionsByPriceAndNonce(signer, localTxs, header.BaseFee)
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
+			w.handleCCCResult(result)
+			return
+		}
+	}
+	if len(remoteTxs) > 0 {
+		txs := types.NewTransactionsByPriceAndNonce(signer, remoteTxs, header.BaseFee)
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
+			w.handleCCCResult(result)
+			return
+		}
+	}
+}
+
+func (w *worker) handleCCCResult(res *pipeline.Result) error {
+	if res != nil && res.OverflowingTx != nil {
+		if res.FinalBlock == nil {
+			// first txn overflowed the circuit, skip
+			log.Trace("Circuit capacity limit reached for a single tx", "tx", res.OverflowingTx.Hash().String(),
+				"isL1Message", res.OverflowingTx.IsL1MessageTx(), "reason", res.CCCErr.Error())
+
+			// Store skipped transaction in local db
+			overflowingTrace := res.OverflowingTrace
+			if !w.config.StoreSkippedTxTraces {
+				overflowingTrace = nil
+			}
+			rawdb.WriteSkippedTransaction(w.eth.ChainDb(), res.OverflowingTx, overflowingTrace, res.CCCErr.Error(),
+				w.currentPipeline.Header.Number.Uint64(), nil)
+
+			if overflowingL1MsgTx := res.OverflowingTx.AsL1MessageTx(); overflowingL1MsgTx != nil {
+				rawdb.WriteFirstQueueIndexNotInL2Block(w.eth.ChainDb(), w.currentPipeline.Header.ParentHash, overflowingL1MsgTx.QueueIndex+1)
```

Comment:
Why write to DB here? Is this the only way to ensure that we'll start with the next l1MsgTx in the next pipeline?

Reply:
Yes, correct. Only way that I could come up with to persist that information between pipeline instances.